tests/aggregate_tests.py

```diff
@@ -45,11 +45,29 @@
 # and run in a test suite.
 tests_list = glob.glob('test_*.py')
 
+# A dictionary of test modules that should only run in certain python versions.
+# Carefully consider the impact of only testing these in a given version.
+# test_proxy_use.py: uses a proxy that only runs in Python2.7. TUF's
+# compatibility with proxies is not likely to vary based on the Python version
+# in use, so this is OK for now. See comments in that module.
+# The format here is: if major is included, limit version to that listed here.
+# If minor is included, limit version to that listed here. Skip the test if any
+# such listed constraints don't match the python version currently running.
+VERSION_SPECIFIC_TESTS = {
+    'test_proxy_use': {'major': 2, 'minor': 7}} # Run test only if Python2.7
+
 # Remove '.py' from each filename to allow loadTestsFromNames() (called below)
 # to properly load the file as a module.
 tests_without_extension = []
 for test in tests_list:
   test = test[:-3]
+  if test in VERSION_SPECIFIC_TESTS:
```

Comment:
Since we're dealing with special cases here, this logic must be made more robust. Particularly, the specification check should **fail harshly** if the desired spec makes no sense, e.g. has a `minor` but no `major`, or neither (see test cases below). So, roughly:
```python
testdict = VERSION_SPECIFIC_TESTS[test]
major, minor = testdict.get("major"), testdict.get("minor")
if not major:
  ERROROUT("A version-specific test that requires no major version makes no sense at all!")
  # XXX Do not continue here! Must fix!
if major != sys.version_info.major:
  NOTIFY("Can't run this test on that Python major version")
  continue
if minor != sys.version_info.minor:
  NOTIFY("Can't run this test on that Python minor version")
  continue
# Otherwise, we're good to go!
```

Reply:
The promised test cases are all generated and tested with this snippet. Anytime it ends up in `Clause 3`, take a close look at the result. For example, `None,None (2, 7) Clause 3` means that an entry in the `VERSION_SPECIFIC_TESTS` dict that did not specify `major` and `minor` at all (!!!) ended up getting run on (a simulated) Python 2.7.

```python
import sys
print("Actual Python version is", sys.version_info)
del sys

# Mock `sys.version_info` so we can use the production code (mostly) unchanged
class sys():
  class version_info():
    major = None
    minor = None

VERSION_SPECIFIC_TESTS = {}
sys_versions = []

# Initialize version specific test dict and system versions list
for major in [None, 2, 3]:
  for minor in [None, 5, 6, 7]:
    name = str(major) + "," + str(minor)
    VERSION_SPECIFIC_TESTS[name] = {}
    if major:
      VERSION_SPECIFIC_TESTS[name]["major"] = major
    if minor:
      VERSION_SPECIFIC_TESTS[name]["minor"] = minor
    if major and minor:
      sys_versions.append((major, minor))


def run_test(test, version):
  sys.version_info.major, sys.version_info.minor = version
  for junk in [1]:
    # Start excerpt: https://github.com/theupdateframework/tuf/pull/781/files#diff-6ec6c904b1b852303ff24f29e2058ad8R64
    if test in VERSION_SPECIFIC_TESTS:
      if 'major' in VERSION_SPECIFIC_TESTS[test] \
          and sys.version_info.major != VERSION_SPECIFIC_TESTS[test]['major']:
        print(test, version, "Clause 1")
        continue
      elif 'minor' in VERSION_SPECIFIC_TESTS[test] \
          and sys.version_info.minor != VERSION_SPECIFIC_TESTS[test]['minor']:
        print(test, version, "Clause 2")
        continue
      print(test, version, "Clause 3")
      # End excerpt: https://github.com/theupdateframework/tuf/pull/781/files#diff-6ec6c904b1b852303ff24f29e2058ad8R71



print("Version specific tests:", VERSION_SPECIFIC_TESTS)
print()
print("test   OSver   clause")

for version in sys_versions:
  for test in VERSION_SPECIFIC_TESTS:
    run_test(test, version)
```